pyscf/df/grad/rhf.py

```diff
@@ -241,6 +291,127 @@ def get_int3c(shls_slice=None):
+    a density matrix.
+
+    Args:
+        mf_grad : instance of :class:`Gradients`
+        mol : instance of :class:`gto.Mole`
+        auxmol : instance of :class:`gto.Mole`
+        orbol : list of length nset of ndarrays of shape (nao,*)
+            Contains non-null eigenvectors of density matrices. See
+            docstring for _decompose_1rdm.
+        orbor : list of length nset of ndarrays of shape (nao,*)
+            Contains orbol multiplied by eigenvalues. See docstring for
+            _decompose_1rdm.
+
+    Returns:
+        rhoj : ndarray of shape (nset, naux)
+            Aux-basis densities
+        get_rhok : callable taking 3 integers (i,j,k), i,j<naux, k<nset
+            Returns C-contiguous ndarray of shape
+            (k-j,orbol[i].shape[1],nao) which contains a mixed-basis
+            density matrix
+    '''
+
+    nset = len (orbol)
+    nao, naux = mol.nao, auxmol.nao
+    nbas, nauxbas = mol.nbas, auxmol.nbas
+    ao_loc = mol.ao_loc
+    nocc = [o.shape[-1] for o in orbor]
+
+    int2c = scipy.linalg.cho_factor(auxmol.intor('int2c2e', aosym='s1'))
+    get_int3c_s1 = _int3c_wrapper(mol, auxmol, 'int3c2e', 's1')
+    rhoj = numpy.zeros((nset,naux))
+    f_rhok = lib.H5TmpFile()
+    t1 = (logger.process_clock (), logger.perf_counter ())
+    max_memory = mf_grad.max_memory - lib.current_memory()[0]
+    blksize = max_memory * .5e6/8 / (naux*nao)
+    mol_ao_ranges = balance_partition(ao_loc, blksize)
+    nsteps = len(mol_ao_ranges)
+    t2 = t1
+    for istep, (shl0, shl1, nd) in enumerate(mol_ao_ranges):
+        int3c = get_int3c_s1((0, nbas, shl0, shl1, 0, nauxbas))
+        t2 = logger.timer_debug1 (mf_grad, 'df grad intor (P|mn)', *t2)
+        p0, p1 = ao_loc[shl0], ao_loc[shl1]
+        for i in range(nset):
+            # MRH 05/21/2020: De-vectorize this because array contiguity -> multithread efficiency
+            v = lib.dot(int3c.reshape (nao, -1, order='F').T, orbor[i]).reshape (naux, (p1-p0)*nocc[i])
+            t2 = logger.timer_debug1 (mf_grad, 'df grad einsum (P|mn) u_ni N_i = v_Pmi', *t2)
+            rhoj[i] += numpy.dot (v, orbol[i][p0:p1].ravel ())
+            t2 = logger.timer_debug1 (mf_grad, 'df grad einsum v_Pmi u_mi = rho_P', *t2)
+            v = scipy.linalg.cho_solve(int2c, v)
+            t2 = logger.timer_debug1 (mf_grad, 'df grad cho_solve (P|Q) D_Qmi = v_Pmi', *t2)
+            f_rhok['%s/%s'%(i,istep)] = v.reshape(naux,p1-p0,-1)
+            t2 = logger.timer_debug1 (mf_grad, 'df grad cache D_Pmi (m <-> i transpose upon retrieval)', *t2)
+        int3c = v = None
+    rhoj = scipy.linalg.cho_solve(int2c, rhoj.T).T
+    int2c = None
+    t1 = logger.timer_debug1 (mf_grad, 'df grad vj and vk AO (P|Q) D_Q = (P|mn) D_mn solve', *t1)
+    def get_rhok(set_id, p0, p1):
+        buf = numpy.empty((p1-p0,nocc[set_id],nao))
+        col1 = 0
+        for istep in range(nsteps):
+            dat = f_rhok['%s/%s'%(set_id,istep)][p0:p1]
```

Comment:
The temporary file object `f_rhok` may be closed when exiting `_cho_solve_rhojk`. It needs to be attached to an object that can survive as long as the `get_rhok` function.

Reply:
Does ccbf8de fix this?

Reply:
I was wrong for the first version you implemented. f_rhok was held in the closure of get_rhok. You can keep either version as you like.

Reply:
I prefer the new version, since it's feels like good practice to keep the temporary file potentially accessible as a tag IMO.